app/pages/learn/01_tutorial/04_mastering-the-api/02_invoke/00_methodhandle.md

```diff
@@ -0,0 +1,626 @@
+it with the method handle of any combiner method which will accept arguments starting at any preferred position.
+
+Let's assume that we have a `target` method:
+
+```java
+private static void target(int ignored, int sum, int a, int b) {
+    System.out.printf("%s + %s equals %s and %s is ignored%n", a, b, sum, ignored);
+}
+```
+
+Using `foldArguments` we can pre-process subset of its arguments and insert the resulting value as another
+argument and proceed to execution of the `target` method.
+
+In our example, we have arguments `int a, int b` at the end. We can pre-process any amount of arguments, but they all
+must be at the end. Let's say, we would like to calculate a sum of these two values `a` and `b`, so let's create a method
+for that:
+
+```java
+private static int sum(int a, int b) {
+    return a + b;
+}
+```
+
+Where will the resulting value go exactly? It will be inserted into one of the arguments of our `target` method. It must
+be the argument right before the arguments that we are going to fold, so in our example argument `int sum`. The argument
+reserved for the folding result can't be at any other position. If the `target` method needs to accept more arguments not
+related to this folding logic, they all must go first.
+
+Let's create the method handles and see how we should combine them together:
+
+```java
+MethodHandle targetMethodHandle = lookup.findStatic(Example.class, "target",
+        MethodType.methodType(void.class, int.class, int.class, int.class, int.class));
+MethodHandle combinerMethodHandle = lookup.findStatic(Example.class, "sum",
+        MethodType.methodType(int.class, int.class, int.class));
+MethodHandle preProcessedArguments = MethodHandles.foldArguments(targetMethodHandle, 1, combinerMethodHandle);
+```
+
+The `foldArguments` method accepts:
+- Target method handle, in our case the one pointing to `target` method;
+- `int` number specifying the starting position of arguments related to folding. In our case, the `sum` argument
+  is located at position `1`, so we passed `1`. If we skip this argument, `pos` will default to `0`.
+- Combiner method handle, in our case it's the one pointing to `sum` method.
+
+At the end, we can invoke the resulting method handle and pass all the arguments except `sum` which is going to be
+pre-calculated:
+
+```java
+preProcessedArguments.invokeExact(10000, 1, 2); // outputs: "1 + 2 equals 3 and 10000 is ignored"
+```
+
+It is possible that the combiner method processes values but doesn't return anything. In this case, there is no need
+for a result placeholder in `target` method argument list.
+
+### Filter return value
+Similarly to arguments, we can use an adapter that will apply transformations to the return value.
+
+Let's imagine a situation where we have a method that returns a `String`, and we would like to channel any returned
+value from this method into another method that replaces character `d` with `m` and brings the resulting value to the
```

Comment:
"the uppercase"?
```suggestion
value from this method into another method that replaces character `d` with `m` and brings the resulting value to
```

Reply:
i don't think i have push access to this upstream repo otherwise I would just go in and fix some of this stuff... suggest the following text:

```
that replaces character `d` with `m` and uppercases the resulting value.
```